### PR TITLE
#156 review application page

### LINKED ISSUE #156: Show application page for Reviewer
Display the review page with buttons to allow the reviewer to add a review - for review status ~~'In progress'~~ 'Draft'.
The application page would show all responses with the `Review` option - without opening the modal window yet!
Any previously added `review_response` should be displayed next to the question, as designed in [Miro](https://miro.com/app/board/o9J_kh-SeNA=).
Screen Shot 2020-12-07 at 12 50 40 PM

And showing Reviewer suggestion/comment:
Screen Shot 2020-12-14 at 11 19 56 AM

The sections have a collapse icon and a feature to stick to the top area when expanded. When reaching the next section now that's the one that will get sticked on the wireframe now. Collapsed sections don't get sticked.

### PR DESCRIPTION
**Pull request by nmadruga**
Fixes #156 

~~Based on PR #205, please merge that one before this.~~ **DONE**
Based on PR #219, please merge that one before this.

Requires [back-end PR152](https://github.com/openmsupply/application-manager-server/pull/152) which updates the application_trigger_states view.

## Changes
- Some layout changes to make the Review look as UI designs. Not very precise, just more similar.
- create new hook `useLoadReview` that create a new structure (similar to what is done in `ApplicationOverview` with sections pages and its elements, response and review to be displayed, it uses other hooks:
  - `useLoadApplication` to load the template sections - not all will be part of the review, so we need to first get all sections of the application.
  - `useGetElementsAndResponsesState` - will get all elements in application and its responses, validate the state of each element and return 2 objects one with all elements and another with all sections
  - `useGetReview` - the Graphql hook to retrieve all review_responses (and later more review information) for this review
- Created the component `ReviewSection` to display each section of the application with review buttons. 
- Created a few more constants to be replacing loose strings inside the containers/components.

### CODE REVIEW

File: src/utils/hooks/useLoadReview.tsx [hunk 1]
```diff
@@ -1,34 +1,126 @@
...
 
+  const {
+    error: applicationError,
+    application,
+    templateSections,
+    isApplicationLoaded,
+  } = useLoadApplication({
+    serialNumber,
+  })
+
```
Comment (CarlosNZ) [inline src/utils/hooks/useLoadReview.tsx:33]: Would be good if you could merge the changes from #219 into here (and adjust accordingly) so we can check the whole loading process works correctly. If it's not too much hassle?

File: src/utils/hooks/useLoadReview.tsx [hunk 2]
```diff
@@ -1,34 +1,126 @@
...
   const { data, error: apolloError, loading: apolloLoading } = useGetReviewQuery({
     variables: {
       reviewId,
     },
+    skip: !isApplicationLoaded || !elementsState,
   })
 
+  const getResponse = (code: string) => {
+    if (responsesByCode && responsesByCode[code]) return responsesByCode[code]
+    else return null
```
Comment (CarlosNZ) [inline src/utils/hooks/useLoadReview.tsx:48]:
```suggestion
  const getResponse = (code: string) =>
    responsesByCode && responsesByCode[code] ? responsesByCode[code] : null
```

File: src/utils/hooks/useLoadReview.tsx [hunk 3]
```diff
@@ -1,34 +1,126 @@
...
+import getPageElements from '../helpers/getPageElements'
 
 interface UseLoadReviewProps {
   reviewId: number
+  serialNumber: string
 }
 
-const useLoadReview = ({ reviewId }: UseLoadReviewProps) => {
+const useLoadReview = ({ reviewId, serialNumber }: UseLoadReviewProps) => {
+  const [applicationName, setApplicatioName] = useState<string>('')
```
Comment (CarlosNZ) [inline src/utils/hooks/useLoadReview.tsx:20]:
```suggestion
  const [applicationName, setApplicationName] = useState<string>('')
```
Typo ;)

File: src/utils/hooks/useLoadReview.tsx [hunk 4]
```diff
@@ -1,34 +1,126 @@
...
+    else return null
+  }
+
+  useEffect(() => {
+    if (applicationError) {
+      const error = applicationError as ApolloError
+      setError(error.message)
+      return
+    }
+    if (application) setApplicatioName(application.name)
```
Comment (CarlosNZ) [inline src/utils/hooks/useLoadReview.tsx:57]:
```suggestion
    if (application) setApplicationName(application.name)
```
To match typo above

File: src/utils/hooks/useLoadReview.tsx [hunk 5]
```diff
@@ -1,34 +1,126 @@
...
+    if (responsesError) {
+      setError(responsesError)
+      return
+    }
     if (apolloError) {
       setError(apolloError.message)
       return
     }
-    if (data) {
+    if (data && data.review && data.review.reviewResponses) {
```
Comment (CarlosNZ) [inline src/utils/hooks/useLoadReview.tsx:69]:
```suggestion
    if (data?.review?.reviewResponses) {
```
Optional chaining?

File: src/utils/hooks/useLoadReview.tsx [hunk 6]
```diff
@@ -1,34 +1,126 @@
...
+      // TODO: See if this code can be merged with one in ApplicationOverview - very similar
```
Comment (CarlosNZ) [inline src/utils/hooks/useLoadReview.tsx:70]: Good call!

File: src/utils/hooks/useLoadReview.tsx [hunk 7]
```diff
@@ -1,34 +1,126 @@
...
     }
-  }, [apolloError, data])
+  }, [responsesError, apolloError, data])
 
   return {
+    applicationName,
     loading,
     error,
-    data,
+    reviewSections,
```
Comment (CarlosNZ) [inline src/utils/hooks/useLoadReview.tsx:123]: Should we also return things like `review_status`, or will we just add that as we need it?

File: src/utils/hooks/useLoadReview.tsx [hunk 8]
```diff
@@ -1,34 +1,126 @@
 import { useEffect, useState } from 'react'
-import { useGetReviewQuery } from '../generated/graphql'
+import { ApolloError } from '@apollo/client'
+import {
+  ApplicationResponse,
+  ReviewResponse,
+  TemplateElement,
+  useGetReviewQuery,
```
Comment (CarlosNZ) [inline src/utils/hooks/useLoadReview.tsx:7]: Couple of unused imports here.

File: src/components/Review/ReviewSection.tsx [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  reviewSection: SectionElementStates
+  canEdit: boolean
+}
+
+const ReviewSection: React.FC<ReviewSectionProps> = ({ reviewSection, canEdit }) => {
+  const { section, pages } = reviewSection
+  return (
+    <Segment inverted style={{ backgroundColor: 'WhiteSmoke', margin: '15px 50px 0px' }}>
+      <Header as="h2" content={`${section.title}`} style={{ color: 'Grey' }} />
+      {Object.entries(pages).map(([pageName, elements]) => (
```
Comment (CarlosNZ) [inline src/components/Review/ReviewSection.tsx:17]: Getting the same "unique keys" error as yesterday, so should have key attribute on these.

File: src/components/Review/ReviewSection.tsx [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+                  <Grid.Row>
+                    <Grid.Column width={13}>
+                      {/* // TODO: Replace with SummaryViewWrapper */}
+                      <Header>{element.title}</Header>
+                      {response && <Segment basic>{response?.text}</Segment>}
+                    </Grid.Column>
+                    <Grid.Column width={3}>
+                      {category === TemplateElementCategory.Question &&
+                        isVisible &&
+                        isEditable &&
```
Comment (CarlosNZ) [inline src/components/Review/ReviewSection.tsx:36]: Can probably allow reviews/comments on non-editable responses, right? For example, if a response was auto-generated based on other responses, we'd still want to be able to approve/comment, do you think?

File: src/components/Review/ReviewSection.tsx [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+          </Header>
+          {elements.map(({ element, response, review }) => {
+            const { category, isVisible, isEditable } = element
+            return isVisible ? (
+              <Segment key={`ReviewSection_${element.code}`} style={{ margin: 10 }}>
+                <Grid columns={2} verticalAlign="middle">
+                  <Grid.Row>
+                    <Grid.Column width={13}>
+                      {/* // TODO: Replace with SummaryViewWrapper */}
+                      <Header>{element.title}</Header>
```
Comment (CarlosNZ) [inline src/components/Review/ReviewSection.tsx:30]: This shouldn't be `element.title`, it should be `element.parameters.label`. (Although won't matter once you replace with SummaryViewWrapper) (And won't actually work for labels that are dynamic expressions yet)

File: src/containers/Review/ReviewPageWrapper.tsx [hunk 12]
```diff
@@ -1,36 +1,50 @@
...
+        ))}
+      </Form>
+    </>
   ) : (
-    <div>
-      <p>
-        This is the Review page for Review ID#{reviewId} of Application {serialNumber}
-      </p>
-    </div>
+    <Message error header="Problem to load application for review" />
```
Comment (CarlosNZ) [inline src/containers/Review/ReviewPageWrapper.tsx:47]: Replace with `messages` string?

File: src/containers/Review/ReviewPageWrapper.tsx [hunk 13]
```diff
@@ -1,36 +1,50 @@
...
+  })
 
   return error ? (
     <Message error header={strings.ERROR_REVIEW_PAGE} list={[error]} />
   ) : loading ? (
     <Loading />
+  ) : reviewSections ? (
+    <>
+      <Container text textAlign="center">
+        <Label color="blue">{strings.STAGE_PACEHOLDER}</Label>
```
Comment (CarlosNZ) [inline src/containers/Review/ReviewPageWrapper.tsx:27]:
```suggestion
        <Label color="blue">{strings.STAGE_PLACEHOLDER}</Label>
```
Also mentioned in `constants.ts`

File: src/utils/helpers/getPageElements.tsx [hunk 14]
```diff
@@ -7,6 +7,7 @@ interface GetPageElementsProps {
 }
 
 const getPageElements = ({ elementsState, sectionIndex, pageNumber }: GetPageElementsProps) => {
+  // TODO: Should this also remove non-visible elements from the result?
```
Comment (CarlosNZ) [inline src/utils/helpers/getPageElements.tsx:10]: Yes, probably makes sense to filter them here.

File: src/utils/types.ts [hunk 15]
```diff
@@ -148,6 +150,8 @@ interface ResponsesByCode {
   [key: string]: ResponseFull | null
 }
 
+type ReviewDecision = 'Approve' | 'Reject'
```
Comment (CarlosNZ) [inline src/utils/types.ts:153]: Should these be defined by database enum?

**Review by CarlosNZ — CHANGES_REQUESTED** [review]
Looking more like a real app every day! :) 

Nice work, most of my comments just typos or small improvements.

File: src/utils/types.ts (hunk 15 above)
Comment (nmadruga) [reply src/utils/types.ts:153]: Yes. I've changed that on the next PR: #221.

File: src/utils/helpers/getPageElements.tsx (hunk 14 above)
Comment (nmadruga) [reply src/utils/helpers/getPageElements.tsx:10]: Cool. I'll work on that one now.

File: src/utils/hooks/useLoadReview.tsx (hunk 1 above)
Comment (nmadruga) [reply src/utils/hooks/useLoadReview.tsx:33]: Oh, no problem. I can do that, just one moment.

File: src/utils/hooks/useLoadReview.tsx (hunk 3 above)
Comment (nmadruga) [reply src/utils/hooks/useLoadReview.tsx]: Thanks! Reviews really help solving those before it's too late.

File: src/utils/hooks/useLoadReview.tsx (hunk 5 above)
Comment (nmadruga) [reply src/utils/hooks/useLoadReview.tsx:68]: Yes, I guess that's alright! I was thinking what if it doesn't even have the `data` but that would all resolve to `undefined` you are right.

File: src/utils/hooks/useLoadReview.tsx (hunk 7 above)
Comment (nmadruga) [reply src/utils/hooks/useLoadReview.tsx:122]: Yes, I thought about leaving for when it's needed.

File: src/containers/Review/ReviewPageWrapper.tsx (hunk 12 above)
Comment (nmadruga) [reply src/containers/Review/ReviewPageWrapper.tsx:47]: Done!

File: src/components/Review/ReviewSection.tsx (hunk 10 above)
Comment (nmadruga) [reply src/components/Review/ReviewSection.tsx:36]: Ah, yea. Good point! I'll change this now.

File: src/utils/hooks/useLoadReview.tsx (hunk 1 above)
Comment (nmadruga) [reply src/utils/hooks/useLoadReview.tsx:29]: Done!

File: src/components/Review/ReviewSection.tsx (hunk 9 above)
Comment (nmadruga) [reply src/components/Review/ReviewSection.tsx:21]: Done [2fae4a4](https://github.com/openmsupply/application-manager-web-app/pull/218/commits/2fae4a4968c35b987fab2075ced4fab2107e64ba)

File: src/components/Review/ReviewSection.tsx (hunk 11 above)
Comment (nmadruga) [reply src/components/Review/ReviewSection.tsx]: Replaced to be using SummaryViewWrapper now

File: src/utils/hooks/useLoadReview.tsx (hunk 6 above)
Comment (nmadruga) [reply src/utils/hooks/useLoadReview.tsx:69]: Done!

File: src/utils/helpers/getPageElements.tsx (hunk 14 above)
Comment (nmadruga) [reply src/utils/helpers/getPageElements.tsx]: Done! [869017f](https://github.com/openmsupply/application-manager-web-app/pull/218/commits/869017fb10ebc81f9b2f5f1fbaa2462645aa8ceb)

**Review by CarlosNZ — APPROVED** [review]
Great -- very nice tidy-ups.

FYI -- I just tested the trigger checks using the procedure I mentioned [here](https://github.com/openmsupply/application-manager-web-app/pull/219#issuecomment-745718478) and it's all working as it should -- for Review loading and Application loading at least (haven't tried ReviewAssignment yet)

**Comment by CarlosNZ** [comment]
Screen Shot 2020-12-16 at 4 51 59 PM
@nmadruga -- why don't you merge this to `master`, otherwise you'll get all these changes showing up in my PR? (Or wait till you approve that one?)

**Comment by nmadruga** [comment]
> why don't you merge this to master, otherwise you'll get all these changes showing up in my PR? (Or wait till you approve that one?)

@CarlosNZ I have added in the PR description to wait until PR #219 is merged to merge this one.  The reason why is trying to merge with that one is to not show in the diff of the PR (after I have merged that PR into this one).

**Pull request merged**